### PR TITLE
Traversal refactoring

### LINKED ISSUE #26: Conflicts with existing IBrowserPublisher adapters
Currently the REST service directive registers an `IBrowserPublisher` adapter which adapts the context and the request. During traversal ZPublisher uses the most specific adapter to publish the object. We add a browser layer (IGET, IPOST, etc.) to the request to make our adapter as specific as possible. However there are other adapters that are more specific on the context and get chosen instead of our adapter.

One example is the `ImageTraverser` adapter in plone.app.imaging. This adapter is registered for `plone.app.imaging.interfaces.IBaseObject` and `zope.publisher.interfaces.IRequest`. Because `plone.app.imaging.interfaces.IBaseObject` is inheriting from `Products.Archetypes.interfaces.BaseObject` this makes it impossible to register a REST service for `Products.Archetypes.interfaces.IBaseObject` as the `ImageTraverser` adapter is more specific on the context.

This is a general problem and can happen with other `IBrowserPublisher` adapters.

In my opinion we should make sure that a REST service takes always precedence over other `IBrowserPublisher` adapters.

I can think of the following (unverified!) approaches to handle this problem:
1. Remove all interfaces from the request except for our service interface.
2. Provide an interface on the context and remove it after processing. However this would probably cause a write operation for every REST request.
3. Do not use `IBrowserPublisher` adapters for REST services, instead use another way to hack into traversal.

I will examine the proposed approaches next to see what could be a suitable solution. Any other ideas or suggestions are welcome.

/cc @tisto @bloodbare @lukasgraf

### PR DESCRIPTION
**Pull request by buchi**
As already discussed in #26 and probably experienced in #27 the current traversal implementation for REST services does not work in situations when another IPublishTraverse adapter takes precedence. 

Especially registering a service for a very unspecific context like
```
<plone:service
   for="*"
   name="search"
   method="GET"
/>
```
will result in not working REST services because other traversal adapters are more specific for some contexts.

To address this issue i've reimplemented traversal for REST requests the following way:
- Implement a traversal adapter on the site root for REST requests
- Any traversed object is wrapped with a special traversal aware wrapper that handles further traversal
- Services no longer implement IPublishTraverse but are simple browser views.
- To make sure we do not conflict with other browser views, the name of a REST service is prefixed with `'rest_'`. 

Fixes #26 

/cc @tisto @bloodbare @lukasgraf

### CODE REVIEW

File: src/plone/rest/traverse.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+NAME_PREFIX = u'rest_'
+
+
+class RESTTraverse(DefaultPublishTraverse):
+    adapts(IPloneSiteRoot, IAPIRequest)
+
+    def publishTraverse(self, request, name):
+        try:
+            obj = super(RESTTraverse, self).publishTraverse(request, name)
+        except KeyError:
```
Comment (lukasgraf) [inline src/plone/rest/traverse.py]: I noticed that here only `KeyError` is being handled, whereas in the wrapper below it's `except (KeyError, AttributeError)`. Is there a particular reason for this difference?

File: CHANGES.rst [hunk 2]
```diff
@@ -4,7 +4,10 @@ Changelog
 1.0a3 (unreleased)
 ------------------
 
-- Nothing changed yet.
+- Refactor traversal of REST requests by using a traversal adapter on the site
+  root instead of a traversal adapter for each REST service. This prevents
+  REST services from beeing overriden by other traversal adapters.
```
Comment (lukasgraf) [inline CHANGES.rst]: Typo: -> `being`

File: src/plone/rest/traverse.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            obj = adapter.publishTraverse(request, name)
+        except (KeyError, AttributeError):
+            # No object, maybe a named rest service
+            service = queryMultiAdapter((self.context, request),
+                                        name=NAME_PREFIX + name)
+            if service is None:
+                raise
+            return service
+        else:
+            # Wrap object to ensure we hanlde further traversal
```
Comment (lukasgraf) [inline src/plone/rest/traverse.py]: Typo: -> `handle`

**Comment by erral** [comment]
I am trying to make plone.rest and plone.restapi work with Archetypes, and with these changes and the proper set of serializers and converters, Archetypes objects can also be exposed with the API.

**Comment by lukasgraf** [comment]
This looks good to me! :+1: 

I think this is a much cleaner approach than what we had before. Tested locally, works like a charm. Without any other changes I was able to run a test script that identified some weird edge cases before (204 status codes, incorrect 401 and 404 responses, ...) and got sensible status codes and response bodies for all the cases.

File: src/plone/rest/traverse.py (hunk 1 above)
Comment (buchi) [inline src/plone/rest/traverse.py]: Yes, it's by intention. We only want to catch exceptions that might occur when looking up a named service. Traversal is trying to lookup an item by key (`item = folder[key]`) which raises a `KeyError` if the item does not exist. However for non-folderish items a lookup by key results in a `AttributeError: '__getitem__'`. We can safely assume that the portal object is folderish and lookups by key will only raise `KeyErrors`.

Comment (tisto) [inline src/plone/rest/traverse.py]: Would you mind adding a few comments regarding that in the code?

**Comment by tisto** [comment]
@lukasgraf wouldn't it make sense to include your test script with those weird edge cases in the existing test suite? This sounds like a proper set of regression tests to me. :)

**Comment by lukasgraf** [comment]
@tisto yeah, I'm planning to. Since it's an entire matrix of possible resource types, HTTP methods and authentication states, the challenge is to provide meaningful assertion messages in case of test failures, and not have early test failures mask later ones. I'll take a stab at that outside this PR.

**Comment by buchi** [comment]
I've found an issue with the `RESTWrapper` and `MultiHooks` and added a fix for that. On a vanilla Plone site there is a `MultiHook` registered on the portal root. The problem only shows up in combination with virtual hosting where the portal root is traversed twice. The second time the portal is wrapped with our `RESTWrapper`.

Funny thing to know: with virtual hosting traversal hooks of the portal are called twice.

**Comment by tisto** [comment]
@lukasgraf awesome!
